### PR TITLE
Check camelCase names

### PR DESCRIPTION
**Pull request by bezoerb**
modifierName, descendentName and stateName must be written in camel case. 
See https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md

### CODE REVIEW

File: lib/preset-patterns.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ module.exports = {
   suit: {
     componentName: /[A-Z][a-zA-Z]+/,
     componentSelectors: suitSelector,
-    utilitySelectors: /^\.u(?:-[a-z][a-zA-Z0-9]+)+$/,
+    utilitySelectors: /^\.u(?:-[a-z][a-zA-Z0-9]*)+$/,
```
Comment (davidtheclark) [inline lib/preset-patterns.js]: The chance from `+` to `*` here and below seems unrelated ... and why is it necessary? Wouldn't you expect a warning if a class was `MyComponent--` or `.u`? Don't you expect at least one character?

**Comment by davidtheclark** [comment]
This seems like a good change to me. I'd like to run it by the SUIT maintainers. @simonsmith what do you think? And is there anyone else you want to run this by?

**Comment by simonsmith** [comment]
Might be tricky to implement, as you have things like `FlexEmbed-ratio--16by9` which would fail the regex here from what I can tell.

**Comment by davidtheclark** [comment]
Would it work if the first character was changed from `[a-z]` to `[a-z0-9]`?

File: lib/preset-patterns.js (hunk 1 above)
Comment (bezoerb) [inline lib/preset-patterns.js]: the `+` would enforce at least two characters as the first on is covered by `[a-z]`

**Comment by bezoerb** [comment]
i think so. But i think `Component-descendant--modifier` isn't covered at all at the moment. For this we would need something like `'(?:'+OPTIONAL+')*'` as there is a fixed order at the moment.
Are numbers also allowed as first character for the `descendantName` or just for the modifier?

**Comment by bezoerb** [comment]
And what about `Component--modifier-descendent--modifier`

**Comment by simonsmith** [comment]
Does this change also expect two words now? So `Component--descendantName`?

A single word as lower case is also worth checking perhaps, so `Component-descendant`

**Comment by simonsmith** [comment]
> And what about Component--modifier-descendent--modifier

I think that tends to be an anti-pattern so perhaps worry about:

`Component--modifier`
`Component-descendantName`
`Component-descendant`
`Component-descendant--modifer`

Where modifier can be numbers or letters, such as `16by9`

I think covers most examples from the top of my head. We can always run against the SUIT library to check if you like

File: lib/preset-patterns.js (hunk 1 above)
Comment (davidtheclark) [inline lib/preset-patterns.js]: Right! Good call.

**Comment by bezoerb** [comment]
Ok, i'll add some tests and change the modifier.

**Comment by bezoerb** [comment]
Fixes pushed. I'd be happy if you could run the changes against SUIT library

**Comment by simonsmith** [comment]
Works as expected, this looks good to me!

**Comment by davidtheclark** [comment]
So theoretically this might throw a warning where people weren't getting warnings before, which seems like a breaking change. But in fact it's more accurately enforcing the guidelines that it's supposed to enforce. Do you think that this needs to constitute a breaking change & major version? Or can we slip it by as a minor version update, a new feature (that is a better, stricter regexp)? Thoughts?

**Comment by simonsmith** [comment]
Hmm, I think it could be a breaking change depending on how you use the plugin. In my case we fail the build if components fail to lint and to suddenly inherit that error from a minor update might be confusing.

I did run this version against two projects and no unexpected errors cropped up, but I do see people misuse SUIT conventions quite often so we should perhaps allow them to opt in to this change via a major.

**Comment by davidtheclark** [comment]
@simonsmith : Yeah, you're right. No reason to be squeamish about a major release: it's important communication.

**Comment by simonsmith** [comment]
:+1:

File: lib/preset-patterns.js [hunk 2]
```diff
@@ -18,10 +18,10 @@ module.exports = {
  */
 function suitSelector(componentName, presetOptions) {
   var ns = (presetOptions && presetOptions.namespace) ? presetOptions.namespace + '-' : '';
-  var OPTIONAL_PART = '(?:-[a-zA-Z0-9]+)?';
-  var OPTIONAL_MODIFIER = '(?:--[a-zA-Z0-9]+)?';
+  var OPTIONAL_PART = '(?:-[a-z][a-zA-Z0-9]*)?';
+  var OPTIONAL_MODIFIER = '(?:--[a-z0-9][a-zA-Z0-9]*)?';
   var OPTIONAL_ATTRIBUTE = '(?:\\[\\S+\\])?';
-  var OPTIONAL_STATE = '(?:\\.is-[a-zA-Z0-9]+)?';
+  var OPTIONAL_STATE = '(?:\\.is-[a-z][a-zA-Z0-9]*)?';
```
Comment (davidtheclark) [inline lib/preset-patterns.js]: Sorry, just noticed this ... @simonsmith, should descendent parts and states accept initial numbers as well as lowercase letters, like modifier? E.g. are these classes legit: `.Clothes-2pieceSuit`, `.is-4sale`?

Comment (simonsmith) [inline lib/preset-patterns.js]: They're not common, but I don't think they would be an anti-pattern. So perhaps we should let those through

Comment (davidtheclark) [inline lib/preset-patterns.js]: Thanks. I'll make this change before release.

Comment (simonsmith) [inline lib/preset-patterns.js]: Nice one

**Comment by davidtheclark** [comment]
Published in 2.0.0

**Comment by bezoerb** [comment]
\o/
